File: skills/code-optimizer/references/database/python/code.py
import json
from dataclasses import asdict, dataclass
from time import perf_counter


@dataclass(frozen=True)
class Order:
  user_id: str
  total: int


@dataclass
class QueryStats:
  query_count: int = 0
  rows_returned: int = 0
  bytes_returned: int = 0
  elapsed_ms: float = 0.0


class FakeDb:
  def __init__(self, orders: list[Order]) -> None:
    self.orders = orders
    self.stats = QueryStats()

  def reset(self) -> None:
    self.stats = QueryStats()

  def snapshot(self) -> QueryStats:
    return QueryStats(
      query_count=self.stats.query_count,
      rows_returned=self.stats.rows_returned,
      bytes_returned=self.stats.bytes_returned,
      elapsed_ms=self.stats.elapsed_ms,
    )

  def read_orders_for_user(self, user_id: str) -> list[Order]:
    return self._record(
      [order for order in self.orders if order.user_id == user_id]
    )

  def read_orders_for_users(self, user_ids: list[str]) -> list[Order]:
    allowed = set(user_ids)
    return self._record(
      [order for order in self.orders if order.user_id in allowed]
    )

  def _record(self, rows: list[Order]) -> list[Order]:
    start = perf_counter()
    result = [Order(user_id=row.user_id, total=row.total) for row in rows]
    payload = json.dumps([asdict(row) for row in result])
    self.stats.query_count += 1
    self.stats.rows_returned += len(result)
    self.stats.bytes_returned += len(payload.encode("utf-8"))
    self.stats.elapsed_ms += (perf_counter() - start) * 1000
    return result


def totals_bad(db: FakeDb, user_ids: list[str]) -> dict[str, int]:
  totals: dict[str, int] = {}
  for user_id in user_ids:
    orders = db.read_orders_for_user(user_id)
    totals[user_id] = sum(order.total for order in orders)
  return totals


def totals_good(db: FakeDb, user_ids: list[str]) -> dict[str, int]:
  totals = {user_id: 0 for user_id in user_ids}
  for order in db.read_orders_for_users(user_ids):
    totals[order.user_id] += order.total
  return totals

File: skills/code-optimizer/references/database/node/code.ts
const encoder = new TextEncoder();

export type Order = {
  userId: string;
  total: number;
};

export type QueryStats = {
  queryCount: number;
  rowsReturned: number;
  bytesReturned: number;
  elapsedMs: number;
};

export class FakeDb {
  private readonly orders: Order[];
  private stats: QueryStats;

  constructor(orders: Order[]) {
    this.orders = orders;
    this.stats = {
      queryCount: 0,
      rowsReturned: 0,
      bytesReturned: 0,
      elapsedMs: 0,
    };
  }

  reset(): void {
    this.stats = {
      queryCount: 0,
      rowsReturned: 0,
      bytesReturned: 0,
      elapsedMs: 0,
    };
  }

  snapshot(): QueryStats {
    return { ...this.stats };
  }

  readOrdersForUser(userId: string): Order[] {
    return this.record(
      this.orders.filter((order) => order.userId === userId),
    );
  }

  readOrdersForUsers(userIds: string[]): Order[] {
    const allowed = new Set(userIds);
    return this.record(
      this.orders.filter((order) => allowed.has(order.userId)),
    );
  }

  private record(rows: Order[]): Order[] {
    const start = performance.now();
    const result = rows.map((row) => ({ ...row }));
    const payload = JSON.stringify(result);
    this.stats.queryCount += 1;
    this.stats.rowsReturned += result.length;
    this.stats.bytesReturned += encoder.encode(payload).length;
    this.stats.elapsedMs += performance.now() - start;
    return result;
  }
}

export function totalsBad(db: FakeDb, userIds: string[]): Map<string, number> {
  const totals = new Map<string, number>();
  for (const userId of userIds) {
    const orders = db.readOrdersForUser(userId);
    totals.set(
      userId,
      orders.reduce((total, order) => total + order.total, 0),
    );
  }
  return totals;
}

export function totalsGood(db: FakeDb, userIds: string[]): Map<string, number> {
  const totals = new Map<string, number>();
  for (const userId of userIds) {
    totals.set(userId, 0);
  }
  for (const order of db.readOrdersForUsers(userIds)) {
    totals.set(order.userId, (totals.get(order.userId) ?? 0) + order.total);
  }
  return totals;
}
